ls1intum/Artemis
# Add option to checkout solution repository for Haskell exercises.

### Checklist

### Motivation and Context
For Haskell exercises, it sometimes is required to check the student submission by comparing it to the sample solution.
This is already possible now, but only after manual setup of Bamboo.

### Description
This PR adds an option to checkout the solution repository (into a `solution` folder) during testing. The test cases can then execute both the students submission and the sample solution and compare the output (e.g. when testing with random input).

**Note**: It's the instructor's responsibility to ensure that the student's code can not leak the sample solution.

### Steps for Testing
1. Create a Haskell programming exercise and enable the checkout of the solution repository (see screenshots below).
2. Open both the base and solution build plans in Bamboo, and verify that the solution repository is added:
![Peek 2020-10-16 13-46](https://user-images.githubusercontent.com/5084100/96254711-06f33a00-0fb6-11eb-80d2-73b7d4eb6ae7.gif)
3. Check the tasks and verify that the solution repository is checked out into the `solution` path.
![Peek 2020-10-16 13-47](https://user-images.githubusercontent.com/5084100/96254779-20948180-0fb6-11eb-9570-fcf9aa75b53f.gif)
4. Create one Haskell programming exercise and one Java programming exercise without enabling this option and verify that the solution repository is **not** added to the repositories (see step 2) and checkout task (see step 3).

### Test Coverage

* ExerciseService.java: 85%
* programming-exercise.component.ts 95%

### Screenshots
Added an option during exercise creation that is only available for Haskell exercises.
![image](https://user-images.githubusercontent.com/5084100/96253753-7405d000-0fb4-11eb-9013-7ed29f5f002a.png)
**Note**: As we do not persist whether the solution repository should be checked out, we do not show the value of this flag in the `View` and `Edit` dialogs.
